### PR TITLE
Feat/faster base36 implementation

### PR DESCRIPTION
**Pull request by ribasushi**
Ok that will do for now:
```
go-multibase$ go test -count=1 -bench=.
goos: darwin
goarch: amd64
pkg: github.com/multiformats/go-multibase
BenchmarkRoundTrip/base36-8         	  327925	      3683 ns/op
BenchmarkRoundTrip/base36upper-8    	  319947	      3641 ns/op
BenchmarkRoundTrip/base58btc-8      	  386199	      3270 ns/op
BenchmarkRoundTrip/base58flickr-8   	  424201	      2638 ns/op
BenchmarkEncode/base36-8            	  440899	      2785 ns/op
BenchmarkEncode/base36upper-8       	  436546	      2777 ns/op
BenchmarkEncode/base58btc-8         	  755958	      1606 ns/op
BenchmarkEncode/base58flickr-8      	  762080	      1608 ns/op
BenchmarkDecode/base36-8            	 1562305	       769 ns/op
BenchmarkDecode/base36upper-8       	 1558365	       769 ns/op
BenchmarkDecode/base58btc-8         	 1207875	      1057 ns/op
BenchmarkDecode/base58flickr-8      	 1203482	      1041 ns/op
PASS
```

### CODE REVIEW

File: base36.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import (
+	"fmt"
+	"math"
+)
+
+const UcAlphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+const LcAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
+
+var encExpansion int
+var revAlphabet [256]uint64
```
Comment (Stebalien) [inline base36.go]: Why not map to bytes? Or uint8?

File: base36.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+func init() {
+	for i := range revAlphabet {
+		revAlphabet[i] = 36
```
Comment (Stebalien) [inline base36.go]: "invalid" sentinel? Can we const it?

File: base36.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+	for i, c := range UcAlphabet {
+		revAlphabet[byte(c)] = uint64(i)
+		if c > '9' {
+			revAlphabet[byte(c)+32] = uint64(i)
+		}
+	}
+
+	encExpansion = int(
```
Comment (Stebalien) [inline base36.go]: This is going to be really in accurate. It's 2.

Comment (Stebalien) [reply base36.go]: We could use `127/82` as an over-approximation?

File: base36.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	for zcnt < inSize && inBuf[zcnt] == 0 {
+		zcnt++
+	}
+
+	encSize := (inSize - zcnt) * encExpansion
+
+	// Allocate one big buffer up front
+	// Note: pools *DO NOT* help, the overhead of zeroing the val-half (see below)
+	// kills any performance gain to be had
+	outBuf := make([]byte, (zcnt + encSize*2))
```
Comment (Stebalien) [inline base36.go:62]: This should end up getting allocated on the stack, anyways, right?

File: base36.go (hunk 3 above)
Comment (Stebalien) [reply base36.go]: Or we can just use 2.

File: go.mod [hunk 5]
```diff
@@ -0,0 +1,8 @@
+module github.com/multiformats/go-base36
+
+go 1.11
+
+require (
+	github.com/martinlindhe/base36 v1.0.0
```
Comment (Stebalien) [inline go.mod]: Do we still need this?

**Review by Stebalien — COMMENTED** [review]
Tests?

File: base36.go (hunk 1 above)
Comment (ribasushi) [reply base36.go]: Because we use the value [only once doing uint64 math](https://github.com/multiformats/go-base36/blob/11f3b26a8e0e9068d38d6fa365d487dd00c6d057/base36.go#L126-L130), so might as well just keep it like this in memory.

File: base36.go (hunk 2 above)
Comment (ribasushi) [reply base36.go]: Isn't `36` already a const...?
It's being checked here: https://github.com/multiformats/go-base36/blob/11f3b26a8e0e9068d38d6fa365d487dd00c6d057/base36.go#L122

( I also need to add a check back I dropped stupidly 🤦 )

File: base36.go (hunk 3 above)
Comment (ribasushi) [reply base36.go]: It's [what the original code used](https://github.com/mr-tron/base58/blob/master/base58.go#L36) so I replicated it. Doesn't hurt to overallocate.

File: base36.go (hunk 4 above)
Comment (ribasushi) [reply base36.go:62]: Yeah, since we `string()` it in the end it should not escape. But gives me an idea to tripple check the `-m -m` output

File: base36.go (hunk 1 above)
Comment (Stebalien) [reply base36.go]: Yeah, but smaller values might fit better into the CPU cache. If we're going to go there :).

Also, why not do the math with smaller values? I've noticed that intel chips perform better with smaller types (usually).

File: base36.go (hunk 2 above)
Comment (Stebalien) [reply base36.go]: I mean define a constant somewhere. It took me a while to figure out why we were using 36. If it were just something like "invalid", that would be easier to understand.

File: base36.go (hunk 1 above)
Comment (ribasushi) [reply base36.go]: I do not fully understand the algorithm tbh, but with 32bits it overflows ( tests fail ). Let me fiddle with it a bit more...

File: base36.go (hunk 3 above)
Comment (Stebalien) [reply base36.go]: The original code estimates the size of the resulting buffer by first multiplying by 138, then dividing by 100, then adding 1. This just defines a constant multiple of 2.

Comment (Stebalien) [reply base36.go]: That's fine, but we might as well just define it as 2 if we're going to do that.

Comment (ribasushi) [reply base36.go]: Yeah I realized what you said after I implemented it, pushing up in a moment ;)

File: base36.go (hunk 1 above)
Comment (ribasushi) [reply base36.go]: Doesn't make a difference on benches, left it at []byte ( plus much smaller table ):
https://github.com/multiformats/go-base36/blob/438fc2b1a7f50241be57e5d9801441c5050fa9fa/base36.go#L17-L20

File: base36.go (hunk 2 above)
Comment (ribasushi) [reply base36.go]: Duh...
https://github.com/multiformats/go-base36/blob/438fc2b1a7f50241be57e5d9801441c5050fa9fa/base36.go#L18-L24
+
https://github.com/multiformats/go-base36/blob/438fc2b1a7f50241be57e5d9801441c5050fa9fa/base36.go#L118

File: base36.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	// use the second half for the temporary numeric buffer
+	val := outBuf[encSize+zcnt:]
+
+	high = encSize - 1
+	for _, b := range inBuf[zcnt:] {
+		valIdx = encSize - 1
+		for carry = uint(b); valIdx > high || carry != 0; valIdx-- {
+			carry += uint((val[valIdx])) * 256
```
Comment (Stebalien) [inline base36.go]: nit: I'd use uint64 or uint32 to so we don't run into weird overflow issues.

File: base36.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	for _, r := range s {
+		if r > rune(maxDigitOrdinal) || revAlphabet[r] > maxDigitValueB36 {
+			return nil, fmt.Errorf("invalid base36 character (%q)", r)
+		}
+
+		c = uint64(revAlphabet[r])
+
+		for j := len(outi) - 1; j >= 0; j-- {
+			t = uint64(outi[j])*36 + c
+			c = (t >> 32)
```
Comment (Stebalien) [inline base36.go:126]: This is why we need the 64 bits.

**Review by Stebalien — APPROVED** [review]

File: base36.go (hunk 6 above)
Comment (ribasushi) [reply base36.go]: Will fix. I have a very corner case bug it seems, still working on porting the tests. Will merge once this is all resolved.
